File: src/oop-java/ml/metrics/Metrics.java
package ml.metrics;

import java.util.*;

public class Metrics {
    public static double accuracy(double[] y, double[] p){
        int n=y.length, c=0; 
        for(int i=0;i<n;i++) 
            if((int)y[i]==(int)p[i]) c++; 
            return c/(double)n;
    }
    public static double macroF1(double[] y, double[] p){
        Set<Integer> labs=new HashSet<>(); 
        for(double v:y) 
        labs.add((int)v); 
        for(double v:p) 
        labs.add((int)v);
        double s=0; 
        for(int L: labs) s += f1ForLabel(y,p,L); 
        
        return s/labs.size();
    }
    private static double f1ForLabel(double[] y, double[] p, int L){
        double tp=0, fp=0, fn=0;
        for(int i=0;i<y.length;i++){
            boolean yy=((int)y[i])==L, pp=((int)p[i])==L;
            if(pp&&yy) tp++; 
            else if(pp) fp++; 
            else if(yy) fn++;
        }
        double prec = tp==0?0:tp/(tp+fp);
        double rec  = tp==0?0:tp/(tp+fn);
        return (prec+rec)==0?0:2*prec*rec/(prec+rec);
    }
    public static double rmse(double[] y, double[] p){
        double se=0; 
        for(int i=0;i<y.length;i++){ 
            double d=p[i]-y[i]; se+=d*d; 
        } 
        return Math.sqrt(se/y.length);
    }
    public static double r2(double[] y, double[] p){
        double mean=0; 
        for(double v:y) 
        mean+=v; 
        mean/=y.length;
        
        double sst=0, ssr=0; 
        for(int i=0;i<y.length;i++){ 
            double d=y[i]-mean; 
            sst+=d*d; 
            double e=y[i]-p[i]; 
            ssr+=e*e; 
        }
        return 1.0 - ssr/(sst+1e-12);
    }
}

File: src/proc/utils/metrics.cpp
#include <cmath>
#include <set>
#include <algorithm>
#include <vector>

using Vector = std::vector<double>;

double accuracy(const Vector& y_true, const Vector& y_pred) {
    if (y_true.size() != y_pred.size() || y_true.empty()) return 0.0;
    
    int correct = 0;
    for (size_t i = 0; i < y_true.size(); i++) {
        if (std::abs(y_true[i] - y_pred[i]) < 1e-9) {
            correct++;
        }
    }
    return (double)correct / y_true.size();
}

double _f1_for_label(const Vector& y_true, const Vector& y_pred, double label) {
    double tp = 0, fp = 0, fn = 0;
    
    for (size_t i = 0; i < y_true.size(); i++) {
        bool true_pos = (std::abs(y_true[i] - label) < 1e-9);
        bool pred_pos = (std::abs(y_pred[i] - label) < 1e-9);
        
        if (true_pos && pred_pos) tp++;
        else if (!true_pos && pred_pos) fp++;
        else if (true_pos && !pred_pos) fn++;
    }
    
    double precision = tp / (tp + fp + 1e-12);
    double recall = tp / (tp + fn + 1e-12);
    return 2.0 * precision * recall / (precision + recall + 1e-12);
}

double macro_f1(const Vector& y_true, const Vector& y_pred) {
    std::set<double> labels_set(y_true.begin(), y_true.end());
    std::vector<double> labels(labels_set.begin(), labels_set.end());
    
    double sum_f1 = 0.0;
    for (double label : labels) {
        sum_f1 += _f1_for_label(y_true, y_pred, label);
    }
    
    return sum_f1 / labels.size();
}

double rmse(const Vector& y_true, const Vector& y_pred) {
    if (y_true.size() != y_pred.size() || y_true.empty()) return 0.0;
    
    double sum_sq = 0.0;
    for (size_t i = 0; i < y_true.size(); i++) {
        double diff = y_true[i] - y_pred[i];
        sum_sq += diff * diff;
    }
    
    return std::sqrt(sum_sq / y_true.size());
}

double r2_score(const Vector& y_true, const Vector& y_pred) {
    if (y_true.size() != y_pred.size() || y_true.empty()) return 0.0;
    
    // Compute mean
    double mean = 0.0;
    for (double y : y_true) mean += y;
    mean /= y_true.size();
    
    // Total sum of squares
    double sst = 0.0;
    for (double y : y_true) {
        double diff = y - mean;
        sst += diff * diff;
    }
    
    // Residual sum of squares
    double ssr = 0.0;
    for (size_t i = 0; i < y_true.size(); i++) {
        double diff = y_true[i] - y_pred[i];
        ssr += diff * diff;
    }
    
    return 1.0 - ssr / (sst + 1e-12);
}